nii_reader.py

```diff
@@ -1,17 +1,102 @@
+__license__ = "Apache-2.0"
+
+import io
+import os
+import random
+import string
+import tempfile
+import urllib.request
+from typing import Dict, Optional
+
+import nibabel.filebasedimages
+import numpy as np
 from jina import Executor, DocumentArray, requests
+from jina.logging.logger import JinaLogger
+from jina.types.document import _is_datauri
+from monai.config import DtypeLike
 from monai.data.image_reader import NibabelReader
 
 
 class NiiReader(Executor):
-    def __init__(self, *args, **kwargs):
-        super(NiiReader, self).__init__()
-        self.r = NibabelReader()
+    """
+    Load `NIfTI` format images based on `Nibabel` library.
+    """
+
+    def __init__(
+        self,
+        as_closest_canonical: bool = False,
+        nibabel_args: Optional[Dict] = None,
+        dtype: DtypeLike = np.float32,
+        *args,
+        **kwargs,
+    ):
+        """
+        :param as_closest_canonical: if True, load the image as closest to canonical axis format.
+        :param nibabel_args: additional args for `nibabel.load` API. more details about available args:
+            https://github.com/nipy/nibabel/blob/master/nibabel/loadsave.py
+        :param dtype: dtype which the image will be converted to
+        :param args: the *args for Executor
+        :param kwargs: the **kwargs for Executor
+        """
+        super().__init__(*args, **kwargs)
+        self.nibabel_args = nibabel_args or {}
+        self.dtype = dtype or np.float32
+        self.as_closest_canonical = as_closest_canonical or False
+        self.reader = NibabelReader(
+            self.as_closest_canonical, self.dtype, **self.nibabel_args
+        )
+        self.logger = JinaLogger(
+            getattr(self.metas, 'name', self.__class__.__name__)
+        ).logger
 
     @requests
-    def read(self, docs, **kwargs):
+    def load(self, docs: Optional[DocumentArray] = None, **kwargs):
+        """
+        Load the NIfTI images available in the format `.nii` or `.nii.gz` from the Document.uri and store it in the `blob`
+        of `Document` as an ndarray. Check out `https://nifti.nimh.nih.gov/nifti-2/` for the Data Format and shape
+        :param docs: the input Documents with either the image file name or data URI in the `uri` field
+        """
+        if docs is None:
+            return
         for doc in docs:
-            doc.blob = self._extract_array(doc.uri)
+            if doc.uri == '':
+                self.logger.error(f'No uri passed for the Document: {doc.id}')
+                continue
+
+            with tempfile.TemporaryDirectory() as tmpdir:
+                source_fn = (
+                    self._save_uri_to_tmp_file(doc.uri, tmpdir)
+                    if _is_datauri(doc.uri)
+                    else doc.uri
+                )
+                nifti_image = self._extract_array(source_fn)
+                if nifti_image is None:
+                    continue
+                doc.blob = nifti_image
```

Comment:
Can't we just do this?
```suggestion
                doc.blob = nifti_image
```

Reply:
Setting `None` to `blob` throws an error.  
`TypeError: blob is in unsupported type builtins.NoneType`
raised from this file
`C:\Users\kalim\anaconda3\envs\searchenv\lib\site-packages\jina\types\document\__init__.py:654`

Reply:
Got it